### PR TITLE
New Feature: Search by URI from the Command Line

### PR DESCRIPTION
**Pull request by omad**
### Reason for this pull request

It's currently quite difficult to find out whether a given URI is indexed inside an ODC Datacube. It's possible to search by a URI using the Python API, but not from the command line.

While it is possible to search by Dataset ID, we have had some cases where new data has overwritten old data, and we no longer have a record of the Dataset ID in the filesystem or S3 bucket, only in the database, which makes this feature extremely useful.

### Proposed changes

- Add `datacube dataset uri-search` command line feature.

 - [ ] Closes #xxxx
 - [ ] Tests added / passed
 - [X] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/index/_datasets.py [hunk 1]
```diff
@@ -420,7 +420,7 @@ def get_datasets_for_location(self, uri, mode=None):
         Find datasets that exist at the given URI
 
         :param uri: search uri
-        :param str mode: 'exact' or 'prefix'
+        :param str mode: 'exact' or 'prefix' or None to guess
```
Comment (alexgleith) [inline datacube/index/_datasets.py]:
```suggestion
        :param str mode: 'exact', 'prefix' or None (to guess)
```

File: datacube/scripts/dataset.py [hunk 2]
```diff
@@ -462,6 +463,28 @@ def _get_derived_set(index: Index, id_: str) -> Set[Dataset]:
...
+    """
+    Search by dataset locations
+
+    PATHS may be either file paths or URIs
+    """
+    if search_mode == 'guess':
+        # This is what the API expects. I think it should be changed.
+        search_mode = None
+    for path in paths:
+        dss = list(index.datasets.get_datasets_for_location(uri_resolve(base=path), mode=search_mode))
```
Comment (alexgleith) [inline datacube/scripts/dataset.py]: Shorthand `dss` is less readable, but we do it everywhere...

**Review by alexgleith — APPROVED** [review]

File: datacube/scripts/dataset.py (hunk 2 above)
Comment (omad) [reply datacube/scripts/dataset.py]: :(
It can stay consistent for now, and maybe get changed everywhere later.

**Review by uchchwhash — APPROVED** [review]
👍🏽

File: datacube/scripts/dataset.py [hunk 3]
```diff
@@ -462,6 +463,28 @@ def _get_derived_set(index: Index, id_: str) -> Set[Dataset]:
...
+        if not dss:
+            print(f"Not found in index: {path}")
```
Comment (jeremyh) [inline datacube/scripts/dataset.py]: To be friendlier to CLI scripts, maybe send this message to stderr instead of stdout?

**Comment by alexgleith** [comment]
Yo @omad any chance you want to finish this off?

I'm keen to work on getting a release out soon.

**Pull request merged**